### ISSUE TITLE
 False equivalences of SUnits because of normaliseUnitSyntax.

### ISSUE THREAD

**Issue by philderbeast** [adamgundry/uom-plugin#76]
The function `normaliseUnitSyntax` is only used when testing whether two unit singletons are equivalent.

https://github.com/adamgundry/uom-plugin/blob/7c71b224a009c8618140be18fd04f8c9b0b1ee4b/uom-plugin/src/Data/UnitsOfMeasure/Singleton.hs#L75

I added some doctests that point to an error not picked up by any of the existing tests.

``` haskell
> normaliseUnitSyntax ([] :/ [])
-- fromList []

> normaliseUnitSyntax (["m"] :/ [])
-- fromList [("m",1)]

> normaliseUnitSyntax (["m", "m"] :/ [])
-- fromList [("m",2)]

> normaliseUnitSyntax (["m", "m", "m"] :/ [])
-- fromList [("m",3)]
```

No surprises so far.

```haskell
> normaliseUnitSyntax ([] :/ ["m"])
-- fromList [("m",1)]

> normaliseUnitSyntax ([] :/ ["m", "m"])
-- fromList []

> normaliseUnitSyntax ([] :/ ["m", "m", "m"])
-- fromList [("m",1)]

> normaliseUnitSyntax (["m"] :/ ["m"])
-- fromList []

> normaliseUnitSyntax (["m", "m"] :/ ["m"])
-- fromList [("m",-1)]

> normaliseUnitSyntax (["m"] :/ ["m", "m"])
-- fromList [("m",1)]

> normaliseUnitSyntax (["m", "m"] :/ ["m", "m"])
-- fromList [("m",2)]
```

This gives false equivalences with some comparisons.

```haskell
> normaliseUnitSyntax (["m", "m"] :/ []) == normaliseUnitSyntax (["m", "m"] :/ ["m", "m"])
-- True

> normaliseUnitSyntax (["m"] :/ []) == normaliseUnitSyntax (["m"] :/ ["m", "m"])
-- True

**Comment by philderbeast** [comment]
I added some test cases that show this problem.

```haskell
  , testGroup "read equality (avoid false equivalences)"
    [ testCase "1 m/m^2 /= 1 m"  $ read "[u| 1 m/m^2 |]"  @?/= [u| 1 m |]
    , testCase "1 m /= 1 m/m^2"  $ read "[u| 1 m |]"  @?/= [u| 1 m/m^2 |]
    , testCase "1 m/m^2 /= 1 m"  $ read "[u| 1 m/m^2 |]"  @/=? [u| 1 m |]
    , testCase "1 m /= 1 m/m^2"  $ read "[u| 1 m |]"  @/=? [u| 1 m/m^2 |]
    ]
```

I defined `@?/=` and `@/=?` for these tests.

```haskell
assertNEQ
  :: (Show a, Eq a, HasCallStack)
  => a -- ^ The expected value
  -> a -- ^ The actual value
  -> Assertion
assertNEQ expected actual = unless (actual /= expected) (assertFailure msg) where
    msg = "expected: " ++ show actual ++ " /= " ++ show expected

infix 1 @?/=, @/=?

(@?/=) :: (HasCallStack, Eq a, Show a) => a -> a -> Assertion
(@?/=) = flip assertNEQ

(@/=?) :: (HasCallStack, Eq a, Show a) => a -> a -> Assertion
(@/=?) = assertNEQ
```

Before the fix #77, these tests failed with:

```
  read equality (avoid false equivalences)
    1 m/m^2 /= 1 m:                             FAIL
      test-suite-units/Tests.hs:342:
      expected: [u| 1 m |] /= [u| 1 m |]
    1 m /= 1 m/m^2:                             FAIL
      test-suite-units/Tests.hs:343:
      expected: [u| 1 m^-1 |] /= [u| 1 m^-1 |]
    1 m/m^2 /= 1 m:                             FAIL
      test-suite-units/Tests.hs:344:
      expected: [u| 1 m |] /= [u| 1 m |]
    1 m /= 1 m/m^2:                             FAIL
      test-suite-units/Tests.hs:345:
      expected: [u| 1 m^-1 |] /= [u| 1 m^-1 |]
```

**Comment by philderbeast** [comment]
I tweaked the false equivalence tests a bit to expect a parse failure and added these to #77, shown here failing before the fix.

```
  read equality (avoid false equivalences)
    1 m/m^2 /= 1 m:                             FAIL
      test-suite-units/Tests.hs:361:
      No exception!
    1 m /= 1 m/m^2:                             FAIL
      test-suite-units/Tests.hs:361:
      No exception!
```

**Comment by adamgundry** [comment]
Fixed by #77, thanks!
